src/archive/src/main.rs

```diff
@@ -0,0 +1,394 @@
+type UserIndex = StableBTreeMap<VirtualMemory<Memory>, UserIndexKey, ()>;
+
+const GIB: u64 = 1 << 30;
+const WASM_PAGE_SIZE: u64 = 65536;
+/// The maximum number of Wasm pages that we allow to use for the stable storage.
+const MAX_WASM_PAGES: u64 = 8 * GIB / WASM_PAGE_SIZE;
+
+const LOG_INDEX_MEMORY_ID: MemoryId = MemoryId::new(0);
+const LOG_DATA_MEMORY_ID: MemoryId = MemoryId::new(1);
+const USER_INDEX_MEMORY_ID: MemoryId = MemoryId::new(2);
+
+/// Length of the user index key. Changing this value requires a stable memory migration.
+const USER_INDEX_KEY_LENGTH: usize = 24;
+
+thread_local! {
+    /// Static configuration of the archive that init() sets once.
+    static CONFIG: RefCell<ConfigCell> = RefCell::new(ConfigCell::init(
+        config_memory(),
+        ArchiveConfig::default(),
+    ).expect("failed to initialize stable cell"));
+
+    /// Static memory manager to manage the memory available for blocks.
+    static MEMORY_MANAGER: RefCell<MemoryManager<Memory>> = RefCell::new(MemoryManager::init(log_memory()));
+
+    /// Append-only list of encoded blocks stored in stable memory.
+    static LOG: RefCell<StableLog> = with_memory_manager(|memory_manager| {
+        RefCell::new(Log::init(memory_manager.get(LOG_INDEX_MEMORY_ID), memory_manager.get(LOG_DATA_MEMORY_ID)).expect("failed to initialize stable log"))
+    });
+
+    /// Index to efficiently filter entries by user number.
+    static USER_INDEX: RefCell<UserIndex> = with_memory_manager(|memory_manager| {
+        RefCell::new(StableBTreeMap::new(memory_manager.get(USER_INDEX_MEMORY_ID), USER_INDEX_KEY_LENGTH as u32, 0))
+    });
+}
+
+/// Creates a memory region for the configuration stable cell.
+fn config_memory() -> Memory {
+    RestrictedMemory::new(DefaultMemoryImpl::default(), 0..1)
+}
+
+/// Creates a memory region for the append-only block list.
+fn log_memory() -> Memory {
+    RestrictedMemory::new(DefaultMemoryImpl::default(), 1..MAX_WASM_PAGES)
+}
+
+/// A helper function to access the configuration.
+fn with_config<R>(f: impl FnOnce(&ArchiveConfig) -> R) -> R {
+    CONFIG.with(|cell| f(cell.borrow().get()))
+}
+
+/// A helper function to access the memory manager.
+fn with_memory_manager<R>(f: impl FnOnce(&MemoryManager<Memory>) -> R) -> R {
+    MEMORY_MANAGER.with(|cell| f(&*cell.borrow()))
+}
+
+/// A helper function to access the log.
+fn with_log<R>(f: impl FnOnce(&StableLog) -> R) -> R {
+    LOG.with(|cell| f(&*cell.borrow()))
+}
+
+/// A helper function to access the user index.
+fn with_user_index<R>(f: impl FnOnce(&mut UserIndex) -> R) -> R {
+    USER_INDEX.with(|cell| f(&mut *cell.borrow_mut()))
+}
+
+/// Configuration of the archive canister.
+#[derive(Clone, Debug, CandidType, Deserialize)]
+struct ArchiveConfig {
+    /// This canister will accept entries only from this principal.
+    ii_canister: Principal,
+    /// The maximum number of entries returned in a single read canister call.
+    max_entries_per_call: u16,
```

Comment:
is this not reflected in the candid type?

Reply:
You're right, this was missing --> added.